batch-processing/pipeline.go

```diff
@@ -0,0 +1,356 @@
+package main
+
+import (
+	"crypto/md5"
+	"encoding/csv"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const (
+	READ     = "read"
+	DOWNLOAD = "download"
+	CONVERT  = "convert"
+	UPLOAD   = "upload"
+	WRITE    = "write"
+)
+
+var TASKS = []string{DOWNLOAD, CONVERT, UPLOAD}
+
+var (
+	OutputHeader       = []string{"url", "input", "output", "s3url"}
+	FailedOutputHeader = []string{"url"}
+)
+
+var (
+	ErrDuplicateURL    = errors.New("duplicate URL, skipping")
+	ErrImageExistsInS3 = errors.New("image already exists in S3, skipping")
+)
+
+// Pipeline struct
+type Pipeline struct {
+	config        *Config
+	workers       map[string]int
+	maxRetries    uint64    // Max number of retries for downloading and uploading an image
+	urlMap        *sync.Map // this will record urls that are already processed so we don't download again.
+	tmpWorkingDir string
+
+	// channels
+	readOut     chan *ReadOut
+	downloadOut chan *DownloadOut
+	convertOut  chan *ConvertOut
+	uploadOut   chan *UploadOut
+	errOut      chan *ErrOut
+}
+
+func (p *Pipeline) Read(csvReader *csv.Reader, out chan *ReadOut) {
+	// Read the input CSV file
+	// For each line, send the URL to the download channel
+	defer close(out)
+
+	csvReader.FieldsPerRecord = 1
+
+	for {
+		row, err := csvReader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			url := ""
+			if len(row) > 0 {
+				url = row[0]
+			}
+			p.errOut <- &ErrOut{
+				Url: url,
+				Err: err,
+			}
+			continue
+		}
+
+		// check if URL was already read
+		_, ok := p.urlMap.Load(row[0])
+		if ok {
+			p.errOut <- &ErrOut{Err: ErrDuplicateURL}
+			continue
+		}
+
+		p.urlMap.Store(row[0], true)
```

Comment:
There's still a race condition here where two goroutines may both call `Store` at the same time - imagine if both called `Load`, saw there was nothing there, then both called `Store` and continued. The key problem here is that the "Load" and "Store" aren't atomic - they're two separate actions, and other things may happen in parallel or in between the two calls.

There are a couple of ways we'd typically avoid this:
1. Use a lock to lock the map (or some value in it) so that the load and store are done under the same lock (which prevents anyone else from calling load or store while we're doing both). This would prevent two `Load` calls from happening in parallel and both saying "no current item".
2. Use an atomic operation of the form "[compare and swap](https://en.wikipedia.org/wiki/Compare-and-swap)" - with a `sync.Map`, that looks like a [`LoadOrStore`](https://pkg.go.dev/sync#Map.LoadOrStore) - rather than doing a `Load` then a `Store`, you can call `LoadOrStore` to effectively say "Load a value if it's set, otherwise store this value" - because you're only calling one operation (which is designed to be atomic), there's no window where some other thread could call `Load` or `Store` between your two calls to `Load` and `Store`.

Reply:
LoadOrStore seems great, I didn't know that one. I have updated to use it. 

Your explanation of things happening between the two calls makes perfect sense and I hadn't considered that (mainly because I only have 1 goroutine for reads and nothing else is using the map). Nevertheless, it's great to keep in mind and be on the lookout.